yt/frontends/gadget/io.py

```diff
@@ -136,95 +136,95 @@ def _generate_smoothing_length(self, index):
                 end = min(ei, d.size) + offsets[fn]
                 d[si:ei] = hsml[begin:end]
 
-    def _get_smoothing_length(self, data_file, position_dtype, position_shape):
+    def _get_smoothing_length(
+        self, data_file, position_dtype, position_shape, handle=None
+    ):
         ptype = self.ds._sph_ptypes[0]
-        si, ei = data_file.start, data_file.end
         if self.ds.gen_hsmls:
             fn = data_file.filename.replace(".hdf5", ".hsml.hdf5")
+            si, ei = data_file.start, data_file.end
+            with h5py.File(fn, mode="r") as f:
+                ds = f[ptype]["SmoothingLength"][si:ei, ...]
         else:
-            fn = data_file.filename
-        with h5py.File(fn, mode="r") as f:
-            ds = f[ptype]["SmoothingLength"][si:ei, ...]
-            dt = ds.dtype.newbyteorder("N")  # Native
-            if position_dtype is not None and dt < position_dtype:
-                # Sometimes positions are stored in double precision
-                # but smoothing lengths are stored in single precision.
-                # In these cases upcast smoothing length to double precision
-                # to avoid ValueErrors when we pass these arrays to Cython.
-                dt = position_dtype
-            hsml = np.empty(ds.shape, dtype=dt)
-            hsml[:] = ds
-            return hsml
+            # note: if handle is None, _read_file will open it.
+            ds = data_file._read_field(ptype, "SmoothingLength", handle)
+            # note, previously, this method always opened the hdf file and
+            # read f[ptype]["SmoothingLength"] without checking if there
+            # are any particles of ptype. _read_field does include that check
+            # and returns None if there are no particles, so we need to
+            # return an empty array here:
+            if ds is None:
+                return np.empty(0, dtype=position_dtype)
+
+        dt = ds.dtype.newbyteorder("N")  # Native
+        if position_dtype is not None and dt < position_dtype:
```

Comment:
I know this predates you but I am just noticing `dt < position_dtype`.

Reply:
ya, I'm not a fan of how this is done right now. Or at least not a fan of how `position_dtype` gets passed in: `_read_data_file`  does a full read of the coordinates just so it can get the coordinate shape and data type to pass as arguments to `_get_smoothing_length`. Seems like there should be a better way to do that. 

In terms of the checks themselves,based on the comments, the purpose is to make sure the array that's leaving is double precision. So could we just always cast to double precision with `.astype` and skip the check?

Reply:
Let's leave it as-is for now.